### ISSUE TITLE
allow `-target` to accept globs

### ISSUE THREAD

**Issue by tamsky** [hashicorp/terraform#2182]
Suppose one declares several concurrent environments in a single tf config.

To limit cli actions to a single environment at a time, it would be convenient if a way existed, via `-target` to glob-match modules, module_paths, resource_types and resource_names:

`terraform apply -target=aws*.staging*`  # target aws* resource_types

or similarly, for all modules:

`terraform apply -target=*.*.staging*` # target all modules and resource_types

or similarly, for all module_paths:

`terraform apply -target=**.staging*`  # star star = target all modules and submodules

**Comment by scalp42** [comment]
:+1: would love to be able to use as opposed to writing CLIs to wrap Terraform

**Comment by krjensen** [comment]
Hi,

Any plans to fix this?

**Comment by elocnatsirt** [comment]
Was just wondering if this is possible. Would be a very useful feature!

**Comment by joshmullis** [comment]
There are other requests I've seen for "exclude" logic, so it may be useful to have this use regex instead of glob to take care of both requests.

**Comment by mario21ic** [comment]
Where is the issue for "exclude" logic? that it'll resolve me many problems :+1:, maybe like target:
$ terraform plan -exclude=mymodule.resource_not_important

**Comment by kwent** [comment]
Would love this feature as well !

**Comment by pkieszcz** [comment]
Just hit it today... Would love this feature times 1000s

**Comment by ryandeivert** [comment]
++ This feature would be excellent to have

**Comment by ljsommer** [comment]
Wow this is a very old request for a very useful feature.
Great idea, our use case would be to use target to specify the creation of template resources to create fully populated python files for Lambdas which have interpolated their vars from terraform so that we can do local unit and lint testing before interacting with the AWS API.

**Comment by ezalejski** [comment]
I am also voting for this feature...

**Comment by idjaw** [comment]
Would it be correct to relate with this Issue #2253 

Considering they are both trying to achieve somewhat similar results. By supporting also the exclude and this pattern feature, it would make it that much more flexible and powerful.

**Comment by chunyong-lin** [comment]
++ would love to have this feature.

**Comment by kcighon** [comment]
this feature would make things a lot better

**Comment by maxblaze** [comment]
+1

**Comment by tolgaingenc** [comment]
+1

**Comment by cytar** [comment]
+1

**Comment by apparentlymart** [comment]
Please do not post "+1" comments here, since it creates noise for others watching the issue and ultimately doesn't influence our prioritization because we can't actually report on these. Instead, react to the original issue comment with :+1:, which [we _can_ and _do_ report on](https://github.com/hashicorp/terraform/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

**Comment by JnMik** [comment]
Not only the target, the "Data" ressource block as well should be able to target a ressource name "prefix-*"

**Comment by vincentkwok** [comment]
Any schedule on this feature? It helps on managing multiple modules. E.g. modify resources under specific module instead of all resources among modules

**Comment by alexanderwburch** [comment]
+1

**Comment by sdickhoven** [comment]
we are managing 700+ domains (plus records) with terraform in cloudflare. our resource naming is perfect for even just prefix wildcards (if that's easier to implement than full-on globbing):
```
terraform apply -target=cloudflare_record.my_domain_com__*
```

some zones have hundreds of records so it would be incredible tedious to list them all as targets. but all the records are prefixed with the zone name.

so definitely :+1: on this feature.

**Comment by rightaway** [comment]
Second highest voted open issue and still no progress.

**Comment by teamterraform** [comment]
Please do not post "+1" comments here, since it creates noise for others watching the issue and ultimately doesn't influence our prioritization because we can't actually report on these. Instead, react to the original issue comment with +1, which we [_can_ and _do_ report on](https://github.com/hashicorp/terraform/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

**Comment by jomeier** [comment]
Hi,

this request is almost four years old and seems to be rather important. Any progress on that?

Thanks and best regards,

Josef

**Comment by Samlinxia** [comment]
Any updates for this feature?  We use terraform to manage our Kafka Topics.  This feature is will be really useful for us to automate manage our topics since our topics are prefixed by a common cluster name.  This wildcard or RegEx match enable us to only apply to a subset of topics that need to update.

**Comment by schneidexe** [comment]
As a workaround I came up with https://github.com/schneidexe/tfp. Yet, it's a kinda low-tech solution that might not cover all cases ... ;) But maybe it's helpful...

**Comment by phzietsman** [comment]
Also ended up making a little cli app to help with this. https://github.com/phzietsman/tft/releases.

`terraform plan | tft -pattern=aws_s3_bucket -mode=include`

Its crude, but it works.

**Comment by alexlevinta** [comment]
still waiting for this feature years later, either wildcarding or exclude would work

**Comment by lmmattr** [comment]
Returning to this after a while, would being able to "Tag" or group resources be another way to help? I'm thinking of building a canary group for deployments but to make deploying this easier you could prefix all the resources in a canary group with a common name that can then be targeted. 

The same could be achieved with a grouping convention. This could help with any staged deployments as you could deploy certain artifacts in groups with manual triggers between them.

**Comment by alexlevinta** [comment]
> Returning to this after a while, would being able to "Tag" or group resources be another way to help? I'm thinking of building a canary group for deployments but to make deploying this easier you could prefix all the resources in a canary group with a common name that can then be targeted.
> 
> The same could be achieved with a grouping convention. This could help with any staged deployments as you could deploy certain artifacts in groups with manual triggers between them.

It would work for groupings that you define ahead of time, but we generally face this issue when something needs to be changed on some long list of production hosts but not on others. The tagging would resolve staged deployments or groupings that you know before running into issues. The regex target and exclusion would help in many more cases.

**Comment by stacybird** [comment]
It would appear this is still relevant years later.  Did the thumbsup to the initial issue comment.  
Context:  Got here looking for a better way to find terraform plan -target=module.FeatureName.aws_thing_here*.* without having to specify each target.  I see there are cli workarounds, those would have to be integrated with atlantis workflows, which is possible, I'll check them out.  Here's hoping regex becomes a more desirable feature.  Still relevant years later.  Thanks for all you do.

**Comment by rightaway** [comment]
What workarounds are people using? Because this is a very important feature.

**Comment by schneidexe** [comment]
There were at least two implemented/mentioned by @phzietsman and me a few comments above...

**Comment by majormoses** [comment]
Not sure as of what version but you can `terraform plan -target=module.my-module` which will plan for all resources within the module. Not quite the same thing but its at least a step in the right direction. This did solve most of my use cases and can live with until there is more featureful pattern matching.

**Comment by schneidexe** [comment]
That only gets you so far though. Imagine you have a complex setup containing modules for k8s clusters on AWS with lambdas, ELBs, databases, DNS records, etc. and just wanna update your ELBs. Then you need scope it to something like `terraform plan -target=**.aws_elb.*`. In such a case you'd now need to search for all `aws_elb` entries in your state and then add them as individual `-target`s which can become quite cumbersome.

**Comment by majormoses** [comment]
I hear, at work we use CI (jenkins) to help alleviate some of the tedious bits associated. A  `state list` job allows an optional `grep` term to filter. Our `plan` jobs allows a comma separated target list. Not perfect, a bit of CI glue can reduce the pain until its enhanced.

**Comment by schneidexe** [comment]
Yeah, exactly that was also my approach and it's pretty much what my [little wrapper project](https://github.com/schneidexe/tfp) does (see above).

**Comment by whiskeysierra** [comment]
@schneidexe Instead of parsing the human-readable console output of terraform plan, you could render the plan to a file and then use `terraform show` on it. With some `jq` magic, you'd have something less brittle. Although the terraform output is most likely not going to change anytime soon.

**Comment by alahijani** [comment]
Here's my low-tech workaround. Add this function to your `.bashrc`:

```bash
terraform-targets () {
  sed 's/\x1b\[[0-9;]*m//g' | grep -o '# [^( ]* ' | sed " s/^# /-target '/; s/ $/'/; "
}
```

You can use it it in a variety of ways:
```bash
terraform plan | terraform-targets
terraform plan | terraform-targets | grep 'some pattern'
terraform plan | terraform-targets | grep 'some pattern' | xargs -r terraform apply
```

(The `-r` flag to `xargs` means don't run the command if the list is empty.)

**Comment by DanPhonovation** [comment]
@alahijani 
> terraform plan | terraform-targets | grep 'some pattern' | xargs -r terraform apply


When I try your last command for the "yes" prompt, something else gets passed and I get action cancelled. Am I doing something wrong?

**Comment by GeorgiKarapetrov** [comment]
> @alahijani
> 
> > terraform plan | terraform-targets | grep 'some pattern' | xargs -r terraform apply
> 
> When I try your last command for the "yes" prompt, something else gets passed and I get action cancelled. Am I doing something wrong?

I run `terraform plan | terraform-targets | grep 'some pattern' | xargs -r terraform apply -auto-approve` when I am positive it's what I want.

**Comment by suryastef** [comment]
I have another workaround for this, but I think this will apply only for resource that already in state

```
for i in $(terraform state list | grep "some-pattern");
do
  terraform apply -target=$i -auto-approve;
done
```

note: be careful with `-auto-approve`, do `terraform plan` first insteaad  of `terraform apply` directly

**Comment by schneidexe** [comment]
@suryastef keep in mind that will run a separate plan/apply for each resource matching which could take a while esp. if you have e.g. a big remote state file in s3. Also in some cases the state might chnage and resources in the list would be removed as dependecies to others... so it would be beneficial to chain the -target .... -target for each $i and run the command only once.

**Comment by suryastef** [comment]
@schneidexe thank you for your concern, I didn't think that way before
>  Also in some cases the state might chnage

**Comment by wwalker** [comment]
The problem is the lock acquisition and release time for each call, and if they were route53 targets every call could take 3 minutes.  (grumble about route53 (not terraform) taking so long to do ANYTHING...).  so if I wanted to destroy 120 route53 entries
(which I actually do at least weekly), the loop method could take up to 6 hours!!!

> I have another workaround for this, but I think this will apply only for resource that already in state
> 
> ```shell
> for i in $(terraform state list | grep "some-pattern");
> do
>   terraform apply -target=$i -auto-approve;
> done
> ```
> 
> note: be careful with `-auto-approve`, do `terraform plan` first insteaad of `terraform apply` directly

**Comment by rubber-ant** [comment]
+1

**Comment by McHill007** [comment]
+1

**Comment by cusspvz** [comment]
Based on the solution provided by @suryastef at https://github.com/hashicorp/terraform/issues/2182#issuecomment-1241546663, I think this makes more sense as it only makes use of one terraform call. (terraform allows us to define multiple targets at the same **apply** call)

```bash
targets=$(for x in `terraform state list | grep -e "azurerm_key_vault*.*"`; do echo "-target=$x "; done);
echo $targets | xargs terraform apply -var-file=... # add here your own options;
```

**Comment by schneidexe** [comment]
Unfort. that does not cover for targets that were not yet deployed and hence do not yet exist in the state...

**Comment by cusspvz** [comment]
@schneidexe figured that out after commenting it in here. 😬

**Comment by baurmatt** [comment]
Getting the approval prompt with xargs can be accomplished with `--open-tty`:

```bash
terraform plan | terraform-targets | grep 'some pattern' | xargs -r --open-tty terraform apply
```

**Comment by silopolis** [comment]
8 years and a half later... I hardly believe such a widespread need remains unconsidered! :|

Adding to the list of alternatives: https://github.com/future-architect/tftarget

**Comment by DevOpsJon** [comment]
Wow almost a decade of terraform pain and this isn't done yet. Yeah, so Pulumi then...

**Comment by rogerioefonseca** [comment]
Hmmmm.... still not there.
It would clean up my code in a really fashion way!

**Comment by rifuller** [comment]
I wanted to use this as a workaround for the lack of support for dynamically configuring providers per-resource-instance. Essentially, provision the dependency explicitly in one apply pass, fetch the resources' properties manually, and then generating the provider blocks. I am again disappointed in Terraform.

**Comment by vanvuvuong** [comment]
I know this issue hadn't been resolved for long time, but maybe this will help someone. This is another low-level option for my case, `terraform apply` for multiple target based on `*tf` file

```
get_target() {
    # $1: pattern of target resource you want to filter(based on your terraform files), e.g: aws_subnet
    # $2: prefix that you want to add to -target, e.g: module.test.
    # result: -target module.test.aws_subnet.subnet_name
    pattern=$1
    prefix=$2
    apply_targets=""
    for target in $(grep -hRe "^resource.*$pattern" | sed -e 's/resource "//g' | sed 's/" {//g' | sed 's/" "/./g'); do apply_targets="$apply_targets -target $prefix$target"; done
}
get_target "aws_subnet" "module.test." | xargs -r terraform apply -auto-approve
```

**Comment by mchigit** [comment]
Why is there still no progress on this issue? It's been reported since 2015, almost a decade later and still no updates? Clearly people want this feature based on the number of comments on this thread.

**Comment by Positronico** [comment]
That's what I came up with, and it's helping me a lot. I'm using regex on resource names from the tf source and state files. Just added an exclude option to skip some of the previously selected targets for finer control.

Single python script, easy to integrate in a pipeline.

https://github.com/Positronico/terratarget

**Comment by kimllee** [comment]
Thank you @Positronico !

Looks like it can't handle the following context : 

```
terratarget  plan -var-file values.tfvars -target="google_logging_project_sink.*"

Warning: No targets matched the wildcard patterns.

Planned command: terraform plan -var-file values.tfvars

Do you want to execute this command? [y/N]: N
Execution aborted by the user.
```

**Comment by augustovictor** [comment]
Here's how I do it (i'm on macos): 
```shell
terraform state list | grep 'module.some-resource-prefix' | awk '{print "--target="$0}' | xargs | xargs -o terraform plan --destroy
```
`-o`: enables tty to allow for interactive usage (and input `yes`, for example)

ps: can also use with `grep -E '<regex_here>'`;

**Comment by atheiman** [comment]
a decade later and I still need this :(
